src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,435 @@
+		k_panic();
+	}
+
+	tr_info(&ll_tr, "task complete %p %pU", task, task->uid);
+	tr_info(&ll_tr, "num_tasks %d total_num_tasks %d",
+		sch->n_tasks, atomic_read(&sch->ll_domain->total_num_tasks));
+
+	/*
+	 * If this is the last task, domain_unregister() won't return. It is
+	 * important to decrement the task counter last before aborting the
+	 * thread.
+	 */
+	domain_unregister(sch->ll_domain, task, --sch->n_tasks);
+}
+
+/* The caller must hold the lock and possibly disable interrupts */
+static void zephyr_ll_task_insert_unlocked(struct zephyr_ll *sch, struct task *task)
+{
+	struct task *task_iter;
+	struct list_item *list;
+
+	task->state = SOF_TASK_STATE_QUEUED;
+
+	/*
+	 * Tasks are added into the list from highest to lowest priority. This
+	 * way they can then be run in the same order. Tasks with the same
+	 * priority are served on a first-come-first-serve basis
+	 */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+		if (task->priority < task_iter->priority) {
+			list_item_append(&task->list, &task_iter->list);
+			break;
+		}
+	}
+
+	/*
+	 * If the task has not been added, it means that it has the lowest
+	 * priority and should be added at the end of the list
+	 */
+	if (list == &sch->tasks)
+		list_item_append(&task->list, &sch->tasks);
+}
+
+/*
+ * Task state machine:
+ * INIT:	initialized
+ * QUEUED:	inserted into the scheduler queue
+ * PENDING:	the scheduler is running, the task is moved to a temporary list
+ * RUNNING:	the scheduler is running, the task is being executed
+ * CANCEL:	anywhere, where the lock is not held, tasks can be cancelled
+ *		asynchronously from an arbitrary context
+ * other:	never set, shouldn't be used, but RESCHEDULE and COMPLETED are
+ *		returned by task's .run function, they are assigned to a
+ *		temporary state, but not to the task .state field
+ */
+
+static void zephyr_ll_run(void *data)
+{
+	struct zephyr_ll *sch = data;
+	struct task *task;
+	struct list_item *list, *tmp, head;
+	uint32_t flags;
+
+	list_init(&head);
+
+	spin_lock_irq(&sch->lock, flags);
+
+	/* Move all pending tasks to a temporary local list */
+	list_for_item_safe(list, tmp, &sch->tasks) {
+		struct comp_dev *sched_comp;
+		struct zephyr_ll_pdata *pdata;
+
+		task = container_of(list, struct task, list);
+		pdata = task->priv_data;
+
+		if (task->state == SOF_TASK_STATE_CANCEL) {
+			zephyr_ll_task_done(sch, task);
+			continue;
+		}
+
+		if (!domain_is_pending(sch->ll_domain, task, &sched_comp))
+			continue;
+
+		/*
+		 * Do not adjust .n_tasks, it's only decremented if the task is
+		 * removed via zephyr_ll_task_done()
+		 */
+		list_item_del(&task->list);
+		list_item_append(&task->list, &head);
+		pdata->run = true;
+
+		/*
+		 * In principle this is optional, it can also be overwritten if
+		 * the task is cancelled, which is an intrinsic race, that
+		 * cannot be prevented as long as we want to run tasks unlocked.
+		 */
+		task->state = SOF_TASK_STATE_PENDING;
```

Comment:
if this is optional then lets remove this state change here.

Reply:
ok